File: 4_median_of_two_arrays/median.go
package medianoftwoarrays

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	i, j, k := 0, 0, 0
	n, m := len(nums1), len(nums2)
	r := n + m
	res := make([]int, r)

	for i < n && j < m {
		if nums1[i] < nums2[j] {
			res[k] = nums1[i]
			i++
		} else {
			res[k] = nums2[j]
			j++
		}
		k++
	}

	for i < n {
		res[k] = nums1[i]
		i++
		k++
	}

	for j < m {
		res[k] = nums2[j]
		j++
		k++
	}

	q := r / 2

	if r%2 == 0 {
		return float64(res[int(q)]+res[int(q)-1]) / 2.0
	}

	return float64(res[q])
}

File: 4_median_of_two_arrays/median.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
const findMedianSortedArrays = (nums1, nums2) => {
  const n = nums1.length;
  const m = nums2.length;

  let i = 0;
  let j = 0;

  const res = [];

  while (i < n && j < m) {
    if (nums1[i] < nums2[j]) {
      res.push(nums1[i]);
      i++;
    } else {
      res.push(nums2[j]);
      j++;
    }
  }

  while (i < n) {
    res.push(nums1[i]);
    i++;
  }

  while (j < m) {
    res.push(nums2[j]);
    j++;
  }

  const q = Math.floor(res.length / 2);

  if (res.length % 2 == 0) {
    return (res[q] + res[q - 1]) / 2;
  }

  return res[q];
};
